### PR TITLE
refactor : ECOM-3-Login-

### PR DESCRIPTION
**Pull request by YasmeenGad**
+ execute login API using clean arch with retrofit
login

### CODE REVIEW

File: lib/core/networking/error/ErrorModel.dart
Comment (Marco-Nagy) [inline lib/core/networking/error/ErrorModel.dart:1]: remove code attribute and refactor error model

Comment (Marco-Nagy) [inline lib/core/networking/error/ErrorModel.dart:1]: remove code attribute and refactor error model

File: lib/core/utils/widgets/custom_toast.dart [hunk 1]
```diff
@@ -0,0 +1,53 @@
+import 'package:flowery_e_commerce/core/styles/colors/my_colors.dart';
+import 'package:flutter/material.dart';
+import 'package:fluttertoast/fluttertoast.dart';
+
+class CustomToast {
+  static void showToast({
+    required String message,
+    required Color backgroundColor,
+    required ToastGravity gravity,
+    required IconData icon,
```
Comment (Marco-Nagy) [inline lib/core/utils/widgets/custom_toast.dart]: Plese refactor this line to replace IconData with Widget to can use animated image in future like lotti files or image assets

File: lib/features/auth/data/repositories/auth_repo_impl.dart [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+@Injectable(as: AuthRepository)
+class AuthRepositoryImpl implements AuthRepository {
+  final AuthOnlineDataSource _onlineDataSource;
+
+  @factoryMethod
+  AuthRepositoryImpl(this._onlineDataSource);
+
+  @override
+  Future<ApiResult<LoginResponseEntity>> login(
```
Comment (Marco-Nagy) [inline lib/features/auth/data/repositories/auth_repo_impl.dart]: Refactor ApiResult class to DataResult to can use it at online data source and offline data source

File: lib/features/auth/domain/contracts/auth_repo.dart
Comment (Marco-Nagy) [inline lib/features/auth/domain/contracts/auth_repo.dart:1]: Refactor ApiResult class to DataResult to can use it at online data source and offline data source

**Review by Marco-Nagy — CHANGES_REQUESTED** [review]
please refactor code with my asked comments

and  Over all you Did a Good Job 
Apricate to be to gather on Teem 2 🙌

**Review by somaya-mahmoud — APPROVED** [review]

File: lib/features/auth/data/data_sources/contracts/auth_online_data_source.dart [hunk 3]
```diff
@@ -0,0 +1,7 @@
+import 'package:flowery_e_commerce/core/networking/common/api_result.dart';
+import 'package:flowery_e_commerce/features/auth/domain/entities/request/login_request_entity.dart';
+import 'package:flowery_e_commerce/features/auth/domain/entities/response/login_response_entity.dart';
+
+abstract interface class AuthOnlineDataSource {
+  Future<ApiResult<LoginResponseEntity>> login(LoginRequestEntity request);
+}
```
Comment (somaya-mahmoud) [inline lib/features/auth/data/data_sources/contracts/auth_online_data_source.dart]: abstract class AuthOnlineDatasource without interface .

File: lib/features/auth/data/data_sources/contracts/offline_data_source.dart [hunk 4]
```diff
@@ -0,0 +1,5 @@
+abstract interface class OfflineDataSource {
+  Future<void> cacheToken(String token);
+
+  Future<String?> getToken();
+}
```
Comment (somaya-mahmoud) [inline lib/features/auth/data/data_sources/contracts/offline_data_source.dart:5]: abstract class also here .

**Review by somaya-mahmoud — COMMENTED** [review]
please remove interface keyword from AuthOnlineDataSource and OfflineDataSource classes , thanks for your effort in login code .

**Review by somaya-mahmoud — COMMENTED** [review]
please remove interface keyword from AuthOnlineDatasource , OfflineDataSource , thanks for your effort in login code .

File: lib/core/utils/widgets/custom_toast.dart [hunk 5]
```diff
@@ -7,7 +7,7 @@ class CustomToast {
     required String message,
     required Color backgroundColor,
     required ToastGravity gravity,
-    required IconData icon,
+    required dynamic icon,
```
Comment (Marco-Nagy) [inline lib/core/utils/widgets/custom_toast.dart:10]: Refactor this from dynamic to Widget

to can use any widget like image asset or Lottie file asset as Json file

**Review by Marco-Nagy — CHANGES_REQUESTED** [review]
Plese Refactor Icon Data To widget

File: lib/core/utils/widgets/custom_toast.dart (hunk 5 above)
Comment (YasmeenGad) [reply lib/core/utils/widgets/custom_toast.dart:10]: An error appears as a result of this modification ( when I replace it with Widget )

Comment (Marco-Nagy) [reply lib/core/utils/widgets/custom_toast.dart:10]: If the package can able to Icon Data  and  not any Widget 
I am sorry You can return it to Icon Data 
Best Of Luck 👍

**Review by Marco-Nagy — CHANGES_REQUESTED** [review]
If the package can able to Icon Data and not any Widget
I am sorry You can return it to Icon Data
Best Of Luck 👍

**Review by Marco-Nagy — CHANGES_REQUESTED** [review]
If the package can able to Icon Data and not any Widget
I am sorry You can return it to Icon Data
Best Of Luck 👍

**Review by Marco-Nagy — APPROVED** [review]

**Pull request merged**